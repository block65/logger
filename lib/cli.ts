import { createNamespace } from 'cls-hooked';
import pino from 'pino';
import {
  callerMixin,
  composeMixins,
  createContextMixin,
  MixinFnWithData,
} from './mixins.js';
import { defaultLoggerOptions } from './options.js';
import { PrettyTransportOptions } from './pretty-transport.js';
import { Falsy, Logger } from './types.js';

export interface CreateCliLoggerOptions
  extends Omit<pino.LoggerOptions, 'mixin' | 'prettyPrint' | 'prettifier'> {
  traceCaller?: boolean;
  mixins?: (MixinFnWithData | pino.MixinFn | Falsy)[];
}

/** @private */
let counter = 0;

export function createCliLogger(
  opts: CreateCliLoggerOptions = {},
  destination: number | string = 1,
): Logger {
  const cls = createNamespace(`logger/${counter}`);
  counter += 1;

  const isDevelopment = process.env.NODE_ENV === 'development';

  const { traceCaller = isDevelopment, mixins = [], ...userPinoOpts } = opts;

  const mixin = composeMixins([
    ...mixins,
    createContextMixin(cls),
    traceCaller && callerMixin,
  ]);

  const prettyTransportOptions: PrettyTransportOptions = {
    destination,
  };

  const pinoInstance = pino({
    ...defaultLoggerOptions,
    ...userPinoOpts,
    mixin,
    transport: {
      targets: [
        {
          target: './pretty-transport.js',
          level: 'trace',
          options: prettyTransportOptions,
        },
      ],
    },
  });

  return Object.create(pinoInstance, {
    cls: {
      value: cls,
      configurable: false,
    },
  });
}
