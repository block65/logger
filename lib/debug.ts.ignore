// import debug from 'debug';
//
// // import pkgUp = require('read-pkg-up');
// // const pkg = pkgUp.sync().pkg;
//
// // TODO: work out how browsers can share this lib without the need for `fs`
// // to get the package name
// const pkg = {
//   name: 'colacube',
// };
//
// const prefixedDebug = debug(pkg.name);
// const debugExtend = (ns: string): debug.Debugger => prefixedDebug.extend(ns);
//
// export { debugExtend, prefixedDebug as debug };
//
// // const myLogger = _debug('logger');
// // myLogger('logger: %o', { pkg });
