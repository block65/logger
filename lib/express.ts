import type { RequestHandler } from 'express';
import type { Logger } from './index.js';
import type { AlsContext } from './logger.js';

export function expressLoggerContextMiddleware(
  logger: Logger,
  id: AlsContext['id'],
  context?: AlsContext['context'],
): RequestHandler {
  return (_, res, next) => {
    res.set('x-context-id', id);

    logger.als.run({ id, ...(context && { context }) }, () => {
      next();
    });
  };
}
