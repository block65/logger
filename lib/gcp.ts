import pino from 'pino';
import { isPlainObject } from './utils.js';

type GcpSeverities = 'DEBUG' | 'INFO' | 'WARNING' | 'ERROR' | 'CRITICAL';

export const logLevelGcpSeverityMap = new Map<
  pino.Level | string,
  GcpSeverities
>([
  ['trace', 'DEBUG'],
  ['debug', 'DEBUG'],
  ['info', 'INFO'],
  ['warn', 'WARNING'],
  ['error', 'ERROR'],
  ['fatal', 'CRITICAL'],
]);

export function formatGcpLogObject(details: object): object {
  if (isPlainObject(details)) {
    if ('stack' in details) {
      const { stack, ...rest } = details;
      if (stack) {
        return {
          stack_trace: stack,
          ...rest,
        };
      }
    }

    if (
      ('err' in details && isPlainObject(details.err)) ||
      details.err instanceof Error
    ) {
      const { err, ...rest } = details;
      const { stack, ...restErr } = err;
      if (stack) {
        return {
          stack_trace: stack,
          ...rest,
          ...restErr,
        };
      }
    }
  }
  return details;
}
