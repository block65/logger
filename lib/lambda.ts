import type { Context } from 'aws-lambda';
import type { JsonObject } from 'type-fest';
import type { Logger } from './logger.js';
import { isEmptyObject } from './utils.js';

type LambdaContext = Partial<Context> & Required<Pick<Context, 'awsRequestId'>>;

export function withLambdaLoggerContextWrapper<T>(
  logger: Logger,
  lambdaContext: LambdaContext,
  fn: (...args: any) => Promise<T>,
): Promise<T> {
  const context: JsonObject = {
    // functionName: context.functionName,
    ...(lambdaContext.functionVersion !== '$LATEST' && {
      functionVersion: lambdaContext.functionVersion,
    }),
  };

  return logger.als.run(
    {
      id: lambdaContext.awsRequestId,
      ...(!isEmptyObject(context) && { context }),
    },
    () => {
      return fn();
    },
  );
}
