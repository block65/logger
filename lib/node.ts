import type { RequestHandler } from 'express';
import { serializeError } from 'serialize-error';
import * as pino from 'pino';
import * as pinoHttp from 'pino-http';
import { createNamespace, Namespace } from 'cls-hooked';
import { hostname } from 'os';
import { randomBytes } from 'crypto';
import { IncomingMessage, ServerResponse } from 'http';
import { sep } from 'path';

type Falsy = false | undefined | null;
type MixinFnWithData = (
  data: ReturnType<pino.MixinFn>,
) => ReturnType<pino.MixinFn>;

export type Logger = {
  cls: Namespace;
} & pino.Logger;

interface ClsContext {
  _contextId: string;
  _context?: Record<string, unknown>;
}

type NamespaceContext = {
  id?: unknown;
  _ns_name?: unknown;
} & Partial<ClsContext>;

type ComputePlatform = 'gcp-cloudrun' | 'aws-lambda' | 'aws';

export interface CreateLoggerOptions extends Omit<pino.LoggerOptions, 'mixin'> {
  traceCaller?: boolean;
  platform?: ComputePlatform;
  mixins?: (MixinFnWithData | pino.MixinFn | Falsy)[];
}

const suggestedHostname = hostname();

const defaultLoggerOptions: pino.LoggerOptions = {
  level: 'info',
  serializers: {
    err: serializeError,
  },
  base: {
    pid: process.pid,
    // in the case we get localhost (gcp cloud run), use random bytes instead
    // this helps us debug if we need to isolate calls running in a specific
    // container or instance (one which stays alive between invocations)
    hostname:
      suggestedHostname !== 'localhost'
        ? suggestedHostname
        : `zz${randomBytes(5).toString('hex')}`, // zz indicates synthetic hostname
  },
};

const gcpLevelMap: Record<string, string> = {
  trace: 'DEFAULT',
  debug: 'DEBUG',
  info: 'INFO',
  warn: 'WARNING',
  error: 'ERROR',
  fatal: 'CRITICAL',
};

function getPlatformLoggerOptions(
  platform?: ComputePlatform,
): pino.LoggerOptions {
  switch (platform) {
    case 'gcp-cloudrun':
      return {
        level: 'info',
        // customLevels: {
        //   DEBUG: 100,
        //   INFO: 200,
        //   NOTICE: 300,
        //   WARNING: 400,
        //   ERROR: 500,
        //   CRITICAL: 600,
        //   ALERT: 700,
        //   EMERGENCY: 800,
        // },
        messageKey: 'message',
        formatters: {
          level(levelLabel /* , levelNumber */) {
            return { severity: gcpLevelMap[levelLabel] };
          },
        },
        timestamp: pino.stdTimeFunctions.isoTime,
      };
    case 'aws':
    case 'aws-lambda':
    default:
      return {};
  }
}

let counter = 0;

const isDevelopment = process.env.NODE_ENV === 'development';

function callerMixin(): { caller: string | undefined } {
  return {
    caller: Error()
      .stack?.split('\n')
      .slice(2)
      .filter(
        (s) =>
          // !s.includes(`api${sep}src${sep}app`) &&
          // !s.includes(
          //   `node_modules${sep}express${sep}lib${sep}router`,
          // ) &&
          !s.includes(`node_modules${sep}pino`) &&
          !s.includes(`block65${sep}logger`),
      )?.[0]
      ?.substr(7),
  };
}

function composeMixins(
  mixins: (MixinFnWithData | pino.MixinFn | Falsy)[],
): pino.MixinFn {
  return () =>
    mixins.reduce((accum, mixin) => {
      if (!mixin) {
        return accum;
      }
      return {
        ...accum,
        ...mixin(accum),
      };
    }, {});
}

function createContextMixin(cls: Namespace): pino.MixinFn {
  return (): Partial<ClsContext> => {
    const { active }: { active: NamespaceContext | null } = cls || {};

    // eslint-disable-next-line @typescript-eslint/naming-convention
    const { id, _ns_name, ...clsContext } = active || {};
    return clsContext;
  };
}

export function createLogger(
  opts: CreateLoggerOptions = {},
  stream?: pino.DestinationStream,
  // destination?: pino.DestinationStream,
): Logger {
  const cls = createNamespace(`logger/${counter}`);
  counter += 1;

  const {
    platform,
    traceCaller = isDevelopment,
    mixins = [],
    ...userPinoOpts
  } = opts;

  const platformLoggerOptions = getPlatformLoggerOptions(platform);

  const mixin = composeMixins([
    ...mixins,
    createContextMixin(cls),
    traceCaller && callerMixin,
  ]);

  const pinoInstance = stream
    ? pino(
        {
          ...defaultLoggerOptions,
          ...platformLoggerOptions,
          ...userPinoOpts,
          mixin,
        },
        stream,
      )
    : pino({
        ...defaultLoggerOptions,
        ...platformLoggerOptions,
        ...userPinoOpts,
        mixin,
      });

  return Object.create(pinoInstance, {
    cls: {
      value: cls,
      configurable: false,
    },
  });
}

export function expressLogger(
  namespace: Namespace,
  contextId: ClsContext['_contextId'],
  context?: ClsContext['_context'],
): RequestHandler {
  return (req, res, next): void => {
    res.set('x-context-id', contextId);

    namespace.run(() => {
      namespace.set('_contextId', contextId);
      if (context) {
        namespace.set('_context', context);
      }
      next();
    });
  };
}

export function lambdaLogger<T extends (...args: any[]) => any>(
  namespace: Namespace,
  contextId: ClsContext['_contextId'],
  context?: ClsContext['_context'],
): (next: T) => (fn: T) => ReturnType<T> {
  return (fn) => {
    return namespace.runAndReturn(() => {
      namespace.set('_contextId', contextId);
      if (context) {
        namespace.set('_context', context);
      }
      return fn();
    });
  };
}

export function createHttpLogger(
  logger: pino.Logger,
  opts: Omit<pinoHttp.Options, 'logger' | 'customLogLevel'> = {},
): pinoHttp.HttpLogger {
  return pinoHttp({
    ...opts,
    genReqId: () => ({}),
    logger,
    customSuccessMessage(res: ServerResponse) {
      return `${res.statusCode} ${res.statusMessage}`;
    },
    customErrorMessage(err: Error, res: ServerResponse) {
      return `${res.statusCode} ${res.statusMessage}`;
    },
    customLogLevel(res: ServerResponse, err: Error) {
      if (res.statusCode >= 500 || err) {
        return 'error';
      }
      if (res.statusCode >= 400 && res.statusCode < 500) {
        return 'warn';
      }
      return 'trace';
    },
    reqCustomProps(req: IncomingMessage) {
      return { url: req.url };
    },
  });
}
