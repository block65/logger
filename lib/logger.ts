import Emittery from 'emittery';
import { AsyncLocalStorage } from 'node:async_hooks';
import { WriteStream } from 'node:fs';
import { PassThrough, Writable } from 'node:stream';
import { finished } from 'node:stream/promises';
import { WriteStream as TtyWriteStream } from 'node:tty';
import format from 'quick-format-unescaped';
import { ErrorObject, serializeError } from 'serialize-error';
import Chain from 'stream-chain';
import type { JsonPrimitive } from 'type-fest';
import { asyncLocalStorageProcessor } from './processors/als.js';
import { isPlainObject, safeStringify } from './utils.js';

// we support an extended set of values, as each have a toJSON method and
// corresponding representation, typically a string
type JsonValueExtended =
  | JsonObjectExtended
  | JsonArrayExtended
  | JsonPrimitive
  | Date
  | URL;

type JsonObjectExtended = { [Key in string]?: JsonValueExtended };
type JsonArrayExtended = JsonValueExtended[];

type JsonObjectExtendedWithError = JsonObjectExtended & {
  err?: Error | unknown;
};

interface LoggerOptions {
  destination: Writable;
  level?: Level;
  processors?: Processor[];
  transformer?: Transformer;
  context?: LogDescriptor['ctx'];
}

export type LoggerProcessor<T, R = Partial<T>> =
  | ((this: Logger, log: T) => R)
  | ((this: Logger, log: T) => Promise<R>);
// | ((log: T) => Readable)
// | ((log: T) => Generator<T, R, unknown>);

export type PlainProcessor<T, R = Partial<T>> =
  | ((log: T) => R)
  | ((log: T) => Promise<R>);
// | ((log: T) => Readable)
// | ((log: T) => Generator<T, R, unknown>);

export type Processor<T = LogDescriptor, R = Partial<T>> =
  | LoggerProcessor<T, R>
  | PlainProcessor<T, R>;

export type PlainTransformer = PlainProcessor<LogDescriptor, unknown>;
export type LoggerTransformer = LoggerProcessor<LogDescriptor, unknown>;
export type Transformer = PlainTransformer | LoggerTransformer;

export enum Level {
  Silent = 0,
  Trace = 10,
  Debug = 20,
  Info = 30,
  Warn = 40,
  Error = 50,
  Fatal = 60,
}

export type LevelAsString = Lowercase<keyof typeof Level>;

export interface CreateLoggerOptions {
  destination?: Writable;
  level?: Level;
  processors?: Processor[];
  transformer?: Transformer;
  context?: LogDescriptor['ctx'];
}

export type LogData = {
  [key in string]?: JsonValueExtended | undefined | unknown;
} & {
  err?: ErrorObject;
};

export interface LogDescriptor {
  time: Date;
  level: Level;
  msg?: JsonPrimitive;
  ctx?: LogData;
  data?: LogData;
  err?: Error | unknown;
}

export type LogMethod =
  | ((err: Error | unknown, str?: JsonPrimitive, ...args: unknown[]) => void)
  | ((
      data: JsonObjectExtendedWithError,
      str?: JsonPrimitive,
      ...args: unknown[]
    ) => void)
  | ((str: JsonPrimitive) => void);

interface LogMethods {
  trace: LogMethod;
  debug: LogMethod;
  info: LogMethod;
  warn: LogMethod;
  error: LogMethod;
  fatal: LogMethod;
}

export interface AlsContext {
  id: string;
  context?: JsonObjectExtended;
}

function stringifyIfNotUndefined(val: unknown): string | undefined {
  return typeof val !== 'undefined' ? String(val) : val;
}

function withNullProto<T extends object>(obj: T, ...objs: Partial<T>[]): T {
  return Object.assign(Object.create(null), obj, ...objs);
}

function isPrimitive(value: unknown): value is JsonPrimitive {
  return (
    (typeof value !== 'undefined' &&
      typeof value !== 'object' &&
      typeof value !== 'function') ||
    value === null
  );
}

function isLogDescriptor(log: LogDescriptor | Symbol): log is LogDescriptor {
  return isPlainObject(log);
}

/**
 * Arguments are all set to unknown as we really cant trust the user
 */
function toLogDescriptor(
  level: Level,
  arg1: unknown,
  arg2?: unknown,
  ...args: unknown[]
): LogDescriptor {
  const time = new Date();

  if (arg1 instanceof Error) {
    return {
      time,
      level,
      msg: typeof arg2 === 'string' ? arg2 : arg1.message,
      data: {
        err: serializeError(arg1),
      },
      err: arg1,
    };
  }

  if (isPrimitive(arg1)) {
    return {
      time,
      level,
      msg: format(String(arg1), [arg2, ...args], {
        stringify: safeStringify,
      }),
    };
  }

  if (Array.isArray(arg1)) {
    return {
      time,
      level,
      msg:
        typeof arg2 === 'string'
          ? format(arg2, args, {
              stringify: safeStringify,
            })
          : stringifyIfNotUndefined(arg2),
      data: { ...arg1 },
    };
  }

  if (arg1 && typeof arg1 === 'object') {
    const { err, ...data } = arg1 as JsonObjectExtendedWithError;

    if (err instanceof Error) {
      return {
        time,
        level,
        msg: typeof arg2 === 'string' ? arg2 : err.message || undefined,
        err,
        data: withNullProto({
          err: serializeError(err),
          ...data,
        }),
      };
    }

    return {
      time,
      level,
      msg:
        typeof arg2 === 'string'
          ? format(String(arg2), args, {
              stringify: safeStringify,
            })
          : stringifyIfNotUndefined(arg2),
      data: arg1,
    };
  }

  return {
    time,
    level,
    msg:
      typeof arg2 === 'string'
        ? format(arg2, args, {
            stringify: safeStringify,
          })
        : stringifyIfNotUndefined(arg1),
  };
}

const pipeCleaner = Symbol('pipe-cleaner');

export class Logger implements LogMethods {
  #inputStream = new PassThrough({
    objectMode: true,
    autoDestroy: true,
  });

  public readonly als: AsyncLocalStorage<AlsContext>;

  public level: Level;

  public destination: Writable | WriteStream | TtyWriteStream;

  #emitter = new Emittery<{
    log: LogDescriptor;
    error: unknown;
    flush: undefined;
  }>();

  #processorChain: Chain;

  #context: LogData | undefined;

  public setLevel(level: Level) {
    this.level = level;
  }

  /**
   *
   * @param options {LoggerOptions}
   */
  constructor(options: LoggerOptions) {
    const {
      level = Level.Info,
      transformer,
      processors = [],
      destination,
      context,
    } = options;

    this.als = new AsyncLocalStorage<AlsContext>();

    this.#context = context && withNullProto(context);

    this.level = level;

    // TODO: validate decorators here
    // validate transformer also

    const processorWrapper = (
      processor: Processor<LogDescriptor | Symbol> | Transformer,
    ): Processor<LogDescriptor | Symbol> | Transformer => {
      if (typeof processor === 'function') {
        return async (log: LogDescriptor | Symbol) => {
          // run the processor on anything log descriptory
          if (isLogDescriptor(log)) {
            try {
              return processor.call(this, log);
            } catch (err) {
              this.#emitter.emit('error', err);
              return log;
            }
          }
          // pass the pipe cleaner (or any symbol) straight through
          return log;
        };
      }

      return processor;
    };

    this.#processorChain = Chain.chain([
      ...[
        asyncLocalStorageProcessor,
        ...processors,
        ...(transformer ? [transformer.bind(this)] : []),
      ].map((processor) => processorWrapper(processor)),
    ]);

    this.destination = destination;

    this.destination.on('error', (err) => this.#emitter.emit('error', err));

    const pipeline = this.#inputStream.pipe(this.#processorChain).pipe(
      // ignore the pipe cleaner
      Chain.chain([(obj) => (obj === pipeCleaner ? null : obj)]),
    );

    if (this.destination.writable) {
      pipeline.pipe(this.destination);
    } else {
      this.#emitter.emit('error', new Error('Destination is not writeable'));
    }

    this.destination.on('end', () => {
      pipeline.unpipe(this.destination);
    });

    pipeline.on('error', (err) => this.#emitter.emit('error', err));
  }

  #write(
    level: Level,
    arg1: unknown,
    arg2?: unknown,
    ...args: unknown[]
  ): void {
    const log = Object.freeze(
      withNullProto(toLogDescriptor(level, arg1, arg2, ...args), {
        ...(this.#context && { ctx: this.#context }),
      }),
    );

    if (log.level >= this.level) {
      if (this.#inputStream.writable) {
        this.#inputStream.write(log, (writeErr: Error | null | undefined) => {
          if (writeErr) {
            this.#emitter.emit('error', writeErr);
          } else {
            // no await here
            this.#emitter
              .emit('log', log)
              .catch((err) => this.#emitter.emit('error', err));
          }
        });
      } else {
        this.#emitter
          .emit('error', new Error('Input stream is not writeable'))
          .catch(() => {});
      }
    }
  }

  public child(
    data: JsonObjectExtended,
    options: Pick<LoggerOptions, 'level' | 'context' | 'processors'> = {},
  ) {
    return new Logger({
      level: this.level,
      destination: this.#inputStream,
      processors: [
        function childLoggerProcessor(log) {
          return {
            ...log,
            data: {
              ...log.data,
              ...data,
            },
          };
        },
        ...(options.processors || []),
      ],
      ...options,
    });
  }

  // https://github.com/microsoft/TypeScript/issues/10570
  public trace(
    err: Error | unknown,
    str?: JsonPrimitive,
    ...args: unknown[]
  ): void;
  public trace(
    data: JsonObjectExtendedWithError,
    str?: JsonPrimitive,
    ...args: unknown[]
  ): void;
  public trace(str: JsonPrimitive, ...args: unknown[]): void;
  // @ts-expect-error
  public trace(...args) {
    // @ts-expect-error
    this.#write(Level.Trace, ...args);
  }

  // https://github.com/microsoft/TypeScript/issues/10570
  public debug(
    err: Error | unknown,
    str?: JsonPrimitive,
    ...args: unknown[]
  ): void;
  public debug(
    data: JsonObjectExtendedWithError,
    str?: JsonPrimitive,
    ...args: unknown[]
  ): void;
  public debug(str: JsonPrimitive, ...args: unknown[]): void;
  // @ts-expect-error
  public debug(...args) {
    // @ts-expect-error
    this.#write(Level.Debug, ...args);
  }

  // https://github.com/microsoft/TypeScript/issues/10570
  public info(
    err: Error | unknown,
    str?: JsonPrimitive,
    ...args: unknown[]
  ): void;
  public info(
    data: JsonObjectExtendedWithError,
    str?: JsonPrimitive,
    ...args: unknown[]
  ): void;
  public info(str: JsonPrimitive): void;
  // @ts-expect-error
  public info(...args) {
    // @ts-expect-error
    this.#write(Level.Info, ...args);
  }

  // https://github.com/microsoft/TypeScript/issues/10570
  public warn(
    err: Error | unknown,
    str?: JsonPrimitive,
    ...args: unknown[]
  ): void;
  public warn(
    data: JsonObjectExtendedWithError,
    str?: JsonPrimitive,
    ...args: unknown[]
  ): void;
  public warn(str: JsonPrimitive): void;
  // @ts-expect-error
  public warn(...args) {
    // @ts-expect-error
    this.#write(Level.Warn, ...args);
  }

  // https://github.com/microsoft/TypeScript/issues/10570
  public error(
    err: Error | unknown,
    str?: JsonPrimitive,
    ...args: unknown[]
  ): void;
  public error(
    data: JsonObjectExtendedWithError,
    str?: JsonPrimitive,
    ...args: unknown[]
  ): void;
  public error(str: JsonPrimitive): void;
  // @ts-expect-error
  public error(...args) {
    // @ts-expect-error
    this.#write(Level.Error, ...args);
  }

  public fatal(
    err: Error | unknown,
    str?: JsonPrimitive,
    ...args: unknown[]
  ): void;
  public fatal(
    data: JsonObjectExtendedWithError,
    str?: JsonPrimitive,
    ...args: unknown[]
  ): void;
  public fatal(str: JsonPrimitive): void;
  // @ts-expect-error
  public fatal(...args) {
    // @ts-expect-error
    this.#write(Level.Fatal, ...args);
  }

  public async flush() {
    await new Promise<void>((resolve, reject) => {
      let t: NodeJS.Timeout | undefined;

      const detectPipeCleaner = (obj: LogDescriptor | Symbol) => {
        if (obj === pipeCleaner) {
          this.#processorChain.off('data', detectPipeCleaner);
          this.#processorChain.off('error', reject);
          if (t) {
            clearTimeout(t);
          }

          resolve();
        }
      };

      this.#processorChain.on('data', detectPipeCleaner);
      this.#processorChain.once('error', reject);

      // timeout waiting for the pipe cleaner
      // t = setTimeout(
      //   () =>
      //     reject(new Error('Timed out waiting for input stream to flush')),
      //   2000,
      // );

      if (this.#inputStream.writable) {
        this.#inputStream.write(pipeCleaner);
      } else {
        // reject(new Error('Stream was not writeable'));
        resolve();
      }
    }).catch((err) => this.#emitter.emit('error', err));

    await this.#emitter
      .emit('flush')
      .catch((err) => this.#emitter.emit('error', err));
  }

  public async end() {
    await this.flush();
    this.#inputStream.end();
    await finished(this.destination);
  }

  public on<Name extends 'log' | 'error' | 'flush'>(
    event: Name | Name[],
    fn: (
      eventData: { log: LogDescriptor; error: unknown; flush: undefined }[Name],
    ) => void | Promise<void>,
  ): Emittery.UnsubscribeFn {
    return this.#emitter.on(event, fn);
  }

  public off<Name extends 'log' | 'error' | 'flush'>(
    event: Name | Name[],
    fn: (
      eventData: { log: LogDescriptor; error: unknown; flush: undefined }[Name],
    ) => void | Promise<void>,
  ): void {
    return this.#emitter.off(event, fn);
  }
}
