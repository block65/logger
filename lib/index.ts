import { hostname } from 'node:os';
import { CreateLoggerOptions, Level, LevelAsString, Logger } from './logger.js';
import { callerProcessor } from './processors/caller.js';
import { createCliTransformer } from './transformers/cli.js';
import { createCloudwatchTransformer } from './transformers/cloudwatch.js';
import { gcpTransformer } from './transformers/gcp.js';
import { jsonTransformer } from './transformers/json.js';

export { Level, Logger } from './logger.js';
export type { CreateLoggerOptions, LogDescriptor } from './logger.js';
export { createRedactProcessor } from './processors/redact.js';
export { createCliTransformer } from './transformers/cli.js';
export { jsonTransformer, type JsonLogFormat } from './transformers/json.js';

const stringLogLevelMap = new Map<LevelAsString | string | undefined, Level>([
  ['silent', Level.Silent],
  ['trace', Level.Trace],
  ['debug', Level.Debug],
  ['info', Level.Info],
  ['warn', Level.Warn],
  ['error', Level.Error],
  ['fatal', Level.Fatal],
]);

function internalCreateLogger(
  options: Omit<CreateLoggerOptions, 'transformer'> = {},
): Logger {
  const recommendedProcessors =
    process.env.NODE_ENV === 'development' ? [callerProcessor] : [];

  const level =
    options.level ||
    stringLogLevelMap.get(process.env.LOG_LEVEL?.toLocaleLowerCase()) ||
    (process.env.NODE_ENV === 'development' ? Level.Trace : undefined);

  const destination = options.destination || process.stdout;

  // Lambda
  if (process.env.AWS_LAMBDA_FUNCTION_VERSION) {
    return new Logger({
      level,
      destination,
      processors: [...recommendedProcessors, ...(options.processors || [])],
      ...options,
      transformer: createCloudwatchTransformer(),
    });
  }

  // ECS
  if (process.env.ECS_AVAILABLE_LOGGING_DRIVERS?.includes('awslogs')) {
    return new Logger({
      level,
      destination,
      transformer: createCloudwatchTransformer(),
      processors: [...recommendedProcessors, ...(options.processors || [])],
      context: {
        ...options.context,
        pid: process.pid,
        hostname: hostname(),
      },
      ...options,
    });
  }

  // GCP (Knative)
  if (
    'K_CONFIGURATION' in process.env &&
    'K_SERVICE' in process.env &&
    'K_REVISION' in process.env
  ) {
    return new Logger({
      level,
      destination,
      processors: [...recommendedProcessors, ...(options.processors || [])],
      transformer: gcpTransformer,
      context: {
        ...options.context,
        pid: process.pid,
        hostname: hostname(),
      },
      ...options,
    });
  }

  // TTY or CI
  if (process.env.CI || ('isTTY' in destination && destination.isTTY)) {
    return new Logger({
      level,
      destination,
      transformer: createCliTransformer(),
      ...options,
    });
  }

  return new Logger({
    level,
    destination,
    transformer: jsonTransformer,
    ...options,
  });
}

function trySentry(logger: Logger) {
  import('@sentry/node')
    .then(async (Sentry) => {
      const sentryOptions = Sentry.getCurrentHub().getClient()?.getOptions();
      // if sentry is configured with a DSN, attach a sentry processor
      if (sentryOptions?.dsn) {
        const { attachSentryListener } = await import('./listener/sentry.js');
        attachSentryListener(logger);
      }
    })
    .catch(() => {
      // no sentry module
    });
}

export function createLogger(
  options: Omit<CreateLoggerOptions, 'transformer'> = {},
): Logger {
  const logger = internalCreateLogger(options);
  trySentry(logger);
  return logger;
}
