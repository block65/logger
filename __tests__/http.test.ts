import * as pinoHttp from 'pino-http';
import * as pino from 'pino';
import * as http from 'http';
import { URL } from 'url';
import { createHttpLogger, createLogger } from '../lib/node';
import { expectStream } from './helpers';

const localPort = 18888;
const agent = new http.Agent({
  // @ts-ignore
  localPort, // reproducible tests
});

export async function doHttpStuff(
  httpLogger: pinoHttp.HttpLogger,
  index: number,
  options: {
    statusCode: number;
    url: string;
  },
): Promise<void> {
  const server = http.createServer((req, res) => {
    httpLogger(req, res);
    res.writeHead(options.statusCode);
    res.end('Hello, World!');
  });

  const serverPort = 8888 + index;
  server.listen(serverPort);

  await new Promise((resolve) => {
    const url = new URL(`http://127.0.0.1:${serverPort}${options.url}`);

    http.get(
      {
        agent,
        hostname: url.hostname,
        port: url.port,
        path: url.pathname,
      },
      () => {
        agent.destroy();
        server.close(resolve);
      },
    );
  });
}

export function testHttpLogger(opts: {
  level: string;
}): [pinoHttp.HttpLogger, Promise<pino.LogDescriptor>] {
  const [stream, logPromise] = expectStream();

  const logger = createLogger(opts, stream);
  const httpLogger = createHttpLogger(logger);
  return [httpLogger, logPromise];
}

describe('HTTP', () => {
  test('Errors', async () => {
    const [httpLogger, logPromise] = testHttpLogger({
      level: 'trace',
    });

    const url = '/broken-page';

    await doHttpStuff(httpLogger, 0, {
      statusCode: 500,
      url,
    });

    await expect(logPromise).resolves.toMatchSnapshot();
  });

  test('Warns', async () => {
    const [httpLogger, logPromise] = testHttpLogger({
      level: 'trace',
    });

    const url = '/missing-page';

    await doHttpStuff(httpLogger, 1, {
      statusCode: 404,
      url,
    });

    await expect(logPromise).resolves.toMatchSnapshot();
  });

  test('OKs', async () => {
    const [httpLogger, logPromise] = testHttpLogger({
      level: 'trace',
    });

    const url = '/existing-page';

    await doHttpStuff(httpLogger, 2, {
      statusCode: 200,
      url,
    });

    await expect(logPromise).resolves.toMatchSnapshot();
  });
});
