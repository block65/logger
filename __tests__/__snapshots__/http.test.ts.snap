// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTTP Errors 1`] = `
Object {
  "err": Object {
    "message": "failed with status code 500",
    "stack": "Error: failed with status code 500
    at ServerResponse.onResFinished (/srv/test/node_modules/pino-http/logger.js:73:38)
    at ServerResponse.emit (node:events:406:35)
    at onFinish (node:_http_outgoing:828:10)
    at callback (node:internal/streams/writable:552:21)
    at afterWrite (node:internal/streams/writable:497:5)
    at afterWriteTick (node:internal/streams/writable:484:10)
    at processTicksAndRejections (node:internal/process/task_queues:82:21)",
    "type": "Error",
  },
  "hostname": "testhost",
  "level": 50,
  "msg": "500 Internal Server Error",
  "pid": 1337,
  "req": Object {
    "headers": Object {
      "connection": "close",
      "host": "127.0.0.1:8888",
    },
    "id": Object {},
    "method": "GET",
    "remoteAddress": "::ffff:127.0.0.1",
    "remotePort": 18888,
    "url": "/broken-page",
  },
  "res": Object {
    "headers": Object {},
    "statusCode": 500,
  },
  "responseTime": 0,
  "time": 1234567890000,
  "url": "/broken-page",
}
`;

exports[`HTTP OKs 1`] = `
Object {
  "hostname": "testhost",
  "level": 10,
  "msg": "200 OK",
  "pid": 1337,
  "req": Object {
    "headers": Object {
      "connection": "close",
      "host": "127.0.0.1:8890",
    },
    "id": Object {},
    "method": "GET",
    "remoteAddress": "::ffff:127.0.0.1",
    "remotePort": 18888,
    "url": "/existing-page",
  },
  "res": Object {
    "headers": Object {},
    "statusCode": 200,
  },
  "responseTime": 0,
  "time": 1234567890000,
  "url": "/existing-page",
}
`;

exports[`HTTP Warns 1`] = `
Object {
  "hostname": "testhost",
  "level": 40,
  "msg": "404 Not Found",
  "pid": 1337,
  "req": Object {
    "headers": Object {
      "connection": "close",
      "host": "127.0.0.1:8889",
    },
    "id": Object {},
    "method": "GET",
    "remoteAddress": "::ffff:127.0.0.1",
    "remotePort": 18888,
    "url": "/missing-page",
  },
  "res": Object {
    "headers": Object {},
    "statusCode": 404,
  },
  "responseTime": 0,
  "time": 1234567890000,
  "url": "/missing-page",
}
`;
